gobblin-core/src/main/java/gobblin/instrumented/Instrumentable.java

```diff
@@ -33,4 +38,47 @@
    * @return true if instrumentation is enabled, false otherwise.
    */
   public boolean isInstrumentationEnabled();
+
+  /**
+   * Generate tags that should be added to the {@link gobblin.metrics.MetricContext}.
+   * @return List of tags to add to Metric Context.
+   */
+  public List<Tag<?>> generateTags(State state);
+
+  /**
+   * Generate a new metric context replacing old tags with input tags (only tags with the same keys will be replaced),
+   * and recreate all metrics in this new context.
+   *
+   * <p>
+   *   This method is useful when the state of the extractor changes and the user wants that state change to
+   *   be reflected in the tags of the extractor instrumentation.
+   * </p>
+   *
+   * <p>
+   *   Notice that this method creates a brand new metric context and metrics every time it is called, with the
+   *   associated processing and memory overhead. Use sparingly only for state changes that MUST be visible in
+   *   emitted metrics.
+   * </p>
+   *
+   * @param tags additional tags.
+   */
+  public void switchMetricContext(List<Tag<?>> tags);
```

Comment:
It seems to be both this and the method below are for very special use cases. I don't think they should be part of the `Instrumentable` interface. @chavdar what do you think?